Document nonbonded methods (#916)
openforcefield/openff-toolkit: docs/developing.md

%  '''''''''''''
%  If you define a new class, add new files to autodoc

### Non-bonded methods as implemented in OpenMM

The SMIRNOFF specification describes the contents of a force field, which can be implemented in a number of different ways in different molecular simulation engines.

> The OpenMM implementation provided by the OpenFF Toolkit produces either an `openmm.System` object containing a `openmm.NonbondedForce` object or raises an exception if there is any incompatibility or unimplemented behavior.
> The below table describes these exceptions or which [`NonbondedMethod`](http://docs.openmm.org/latest/userguide/application.html?highlight=ljpme#nonbonded-interactions) is set on the resulting `NonbondedForce`.

Reviewer: ```suggestion
The OpenMM implementation provided by the OpenFF Toolkit either produces an `openmm.System` containing a `openmm.NonbondedForce` object or raises an exception depending on how the non-bonded parameters are specified.
Exceptions are raised when parameters are incompatible with OpenMM (`IncompatibleParameterError`) or otherwise against spec (`SMIRNOFFSpecError`), and also when they are appropriate for the spec but not yet implemented in the toolkit (`SMIRNOFFSpecUnimplementedError`).
This table describes which [`NonbondedMethod`](http://docs.openmm.org/latest/userguide/application.html?highlight=ljpme#nonbonded-interactions) is used in the produced `NonbondedForce`, or else which exception is raised.
```

I had to read this a few times to parse it, this might be a bit clearer. Please correct any incorrect inferences I've made!

> | vdw_method | electrostatics_method | periodic | OpenMM Nonbonded method or exception |
> |------------|-----------------------|----------|--------------------------------------|
> | cutoff     | Coulomb               | True     | `IncompatibleParameterError`
> | cutoff     | Coulomb               | False    | `openmm.NonbondedForce.NoCutoff`
> | cutoff     | reaction-field        | True     | `SMIRNOFFSpecError`
> | cutoff     | reaction-field        | False    | `SMIRNOFFSpecUnimplementedError`
> | cutoff     | PME                   | True     | `openmm.NonbondedForce.PME`
> | cutoff     | PME                   | False    | `openmm.NonbondedForce.NoCutoff`
> | LJPME      | Coulomb               | True     | `IncompatibleParameterError`
> | LJPME      | Coulomb               | False    | `openmm.NonbondedForce.NoCutoff`
> | LJPME      | reaction-field        | True     | `IncompatibleParameterError`
> | LJPME      | reaction-field        | False    | `SMIRNOFFSpecError`
> | LJPME      | PME                   | True     | `openmm.NonbondedForce.LJPME`
> | LJPME      | PME                   | False    | `openmm.NonbondedForce.NoCutoff`

Reviewer: ```suggestion
| vdw_method | electrostatics_method | periodic | OpenMM Nonbonded method or exception |
|------------|-----------------------|----------|--------------------------------------|
| cutoff     | Coulomb               | True     | raises `IncompatibleParameterError`
| cutoff     | Coulomb               | False    | `openmm.NonbondedForce.NoCutoff`
| cutoff     | reaction-field        | True     | raises `SMIRNOFFSpecError`
| cutoff     | reaction-field        | False    | raises `SMIRNOFFSpecUnimplementedError`
| cutoff     | PME                   | True     | `openmm.NonbondedForce.PME`
| cutoff     | PME                   | False    | `openmm.NonbondedForce.NoCutoff`
| LJPME      | Coulomb               | True     | raises `IncompatibleParameterError`
| LJPME      | Coulomb               | False    | `openmm.NonbondedForce.NoCutoff`
| LJPME      | reaction-field        | True     | raises `IncompatibleParameterError`
| LJPME      | reaction-field        | False    | raises `SMIRNOFFSpecError`
| LJPME      | PME                   | True     | `openmm.NonbondedForce.LJPME`
| LJPME      | PME                   | False    | `openmm.NonbondedForce.NoCutoff`
```
This might make it easier to spot the difference between exceptions and non-bonded methods at a glance. Just a thought, don't worry about it if you don't like it.

Notes:

> * The most commonly-used case (including the Parsley line) is row is the sixth row (cutoff vdW, PME electrostatics, periodic topology)

Reviewer 2: This confused me until I realized you meant "sixth row, if you include the header". I think it's clearer to say "fifth".
 
```suggestion
* The most commonly-used case (including the Parsley line) is the fifth row (cutoff vdW, PME electrostatics, periodic topology)
```

Reviewer: I agree with Jeff. The header is the zeroth row 😛

Could also consider marking the row itself; making it bold, or adding an asterisk or highlighting the background or something. A simple asterisk would probably be cleanest for Markdown readability.

Author: Good catch - I more likely forgot how to count than decide a header row counts as a row

* For all cases included a non-periodic topology, `openmm.NonbondedForce.NoCutoff` is currently used.

> * Electrostatics method `reaction-field` can only apply to periodic systems, however it is not currently implemented.

Reviewer: This seems to contradict the table? `SMIRNOFFSpecError` is given for (cutoff, reaction-field, periodic==True) but `SMIRNOFFSpecUnimplementedError` is given for (cutoff, reaction-field, periodic==False)

Author: I think you're right, I must have made a mistake copy-pasting lines - matches the [tests](https://github.com/openforcefield/openff-toolkit/blob/0ab0b025198c2186d068fe296bcc714a8cc8609b/openff/toolkit/tests/test_forcefield.py#L614-L629) as well.

> * LJPME (particle mesh ewald for LJ/vdW itneractions) is not yet fully described in the SMIRNOFF specification.

Reviewer 2: ```suggestion
* LJPME (particle mesh ewald for LJ/vdW interactions) is not yet fully described in the SMIRNOFF specification.
```
